### PR TITLE
feat: retrospective cmo (FPLA-1366)

### PR DESCRIPTION
**Pull request by lewisbirks**
### JIRA link (if applicable) ###

[FPLA-1366](https://tools.hmcts.net/jira/browse/FPLA-1366)

### Change description ###

Added functionality for a judge to mark a CMO during the action CMO event as retrospective.
The functionality for the LA whilst drafting is for the current date to be used.
The date of issue that the judge sets will be repopulated if the CMO already has one associated with it.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: e2e/pages/events/actionCaseManagementOrderEvent.page.js [hunk 1]
```diff
@@ -79,4 +79,9 @@ module.exports = {
     I.waitForElement(this.fields.nextHearingDateList);
     I.selectOption(this.fields.nextHearingDateList, date);
   },
+
+  async enterDateOfIssue(date){
+    I.fillDate(date);
+    await I.retryUntilExists(() => I.click('Continue'), '#allPartiesCustomCMO');
```
Comment (toireasam) [inline e2e/pages/events/actionCaseManagementOrderEvent.page.js]: could move #allPartiesCustomCMO into the fields section of this file, up to you if you'd prefer leave as is - very minor

File: e2e/pages/events/actionCaseManagementOrderEvent.page.js [hunk 2]
```diff
@@ -79,4 +79,9 @@ module.exports = {
     I.waitForElement(this.fields.nextHearingDateList);
     I.selectOption(this.fields.nextHearingDateList, date);
   },
+
+  async enterDateOfIssue(date){
```
Comment (toireasam) [inline e2e/pages/events/actionCaseManagementOrderEvent.page.js]: spacing here just on the bracket

File: e2e/steps_file.js [hunk 3]
```diff
@@ -194,6 +202,16 @@ module.exports = function () {
       await this.completeEvent('Save and continue');
     },
 
+    async fillDate(date, sectionId = 'form'){
```
Comment (toireasam) [inline e2e/steps_file.js]: spacing on bracket again

Comment (toireasam) [reply e2e/steps_file.js]: also does this need to be async?

Comment (lewisbirks) [reply e2e/steps_file.js]: Pulled this in from @tomaszpowroznik branch, not sure on whether it needs to be or not

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/CMODocmosisTemplateDataGenerationService.java [hunk 4]
```diff
@@ -135,6 +137,14 @@
...
     }
 
+    private String getIssuedDate(LocalDate dateOfIssue) {
+        LocalDate date = dateOfIssue;
+        if (date == null) {
+            date = time.now().toLocalDate(); // When LA drafts it will default to current time
+        }
+        return formatLocalDateToString(date, DATE);
+    }
+
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/CMODocmosisTemplateDataGenerationService.java]: how about
```
   private String getIssuedDate(LocalDate dateOfIssue) {
        LocalDate date = defaultIfNull(dateOfIssue, time.now().toLocalDate());
        return formatLocalDateToString(date, DATE);
    }
```

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DateFormatterServiceTest.java [hunk 5]
```diff
@@ -35,28 +38,28 @@
     @Test
     void shouldFormatLocalDateInLongFormat() {
         LocalDate date = createDate();
-        String formattedDate = dateFormatterService.formatLocalDateToString(date, FormatStyle.LONG);
-        assertThat(formattedDate).isEqualTo("1 January 2019");
+        String formattedDate = formatLocalDateToString(date, FormatStyle.LONG);
+        assertThat(formattedDate).isEqualTo(JANUARY_2019);
```
Comment (Gogetter) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/DateFormatterServiceTest.java]: 👀

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java [hunk 6]
```diff
@@ -187,6 +191,6 @@ private void prePopulateHearingDateSelection(List<Element<HearingBooking>> heari
     }
 
     private String formatLocalDateToMediumStyle(LocalDate date) {
-        return dateFormatterService.formatLocalDateToString(date, FormatStyle.MEDIUM);
+        return formatLocalDateToString(date, FormatStyle.MEDIUM);
```
Comment (Gogetter) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DraftCMOService.java]: 👍

**Pull request merged**